site/topics/testing/unittest.rst

```diff
@@ -152,6 +152,122 @@ Writing Unit Tests
+
+    * Test a ``Sphere`` that exists in an arbitrary location with zero ``radius``
+    * Test a ``Sphere`` that exists in an arbitrary location with non-zero ``radius``
+
+* To test all four example cases the same way as the above tests, we would need four separate tests that are nearly identical
+
+.. code-block:: python
+    :linenos:
+
+    import unittest
+
+    class SphereTest(unittest.TestCase):
+
+        # Other test methods not shown for brevity
+
+        def test_diameter_radius_zero_origin_returns_zero(self):
+            sphere = Sphere(Point3D(0, 0, 0), 0)
+            self.assertEqual(0, sphere.diameter())
+
+        def test_diameter_radius_one_origin_returns_two(self):
+            sphere = Sphere(Point3D(0, 0, 0), 1)
+            self.assertEqual(2, sphere.diameter())
+
+        def test_diameter_radius_zero_arbitrary_centre_returns_zero(self):
+            sphere = Sphere(Point3D(1, 1, 1), 0)
+            self.assertEqual(0, sphere.diameter())
+
+        def test_diameter_radius_ten_arbitrary_centre_returns_twenty(self):
+            sphere = Sphere(Point3D(10, 11, 12), 10)
+            self.assertEqual(20, sphere.diameter())
+
+
+* Although there is nothing wrong with the above tests, we can instead, we can make use of ``subTest`` in this scenario
+
+
+.. code-block:: python
+    :linenos:
+    :emphasize-lines: 15, 16
+
+    import unittest
+
+    class SphereTest(unittest.TestCase):
+
+        # Other test methods not shown for brevity
+
+        def test_diameter_various_spheres_returns_correct_diameter(self):
+            cases = [
+                Sphere(Point3D(0, 0, 0), 0),
+                Sphere(Point3D(0, 0, 0), 1),
+                Sphere(Point3D(1, 1, 1), 0),
+                Sphere(Point3D(10, 11, 12), 10.1),
+            ]
+            expecteds = [0, 2, 0, 20.2]
+            for (case, expect) in zip(cases, expecteds):
+                with self.subTest(case=case, expect=expect):
+                    self.assertAlmostEqual(expect, case.diameter(), 5)
+
+
+* In the above example, each test input and expected output were stored in lists
+
+    * I used two separate lists, but there is nothing stopping you from using one list of tuples
+    * The variable names for the lists, ``cases`` and ``expecteds``, were arbitrary and by no means required
+
+* First notice the loop --- there is nothing particularly important for the ``subTest`` here, but the ``zip`` function has not been seen yet
+
+    * This just provides an easy way to loop over data within two lists at the same time
+    * This whole portion could be re-written as follows
+
+        .. code-block:: python
+            :linenos:
+
+            ...
+            for i in range(len(cases)):
+                with self.subTest(case=cases[i], expect=expecteds[i]):
+                    ...
+
+
+* Next, notice ``with self.subTest(case=case, expect=expect)``
+* There is a lot going on here, and much of it we will leave as *important code we need*
```

Comment:
this basically means, *ignore `with`*

Reply:
The fact the kwargs have the same names as the variables may be confusing. Clarifying that the left hand sides are the names and right hand sides are the variables may be prudent.

Further, it's kinda not totally true? `subTest` works as a no-args function. The arguments are for documentation only.

Reply:
Easiest solution was to remove all the keyword args stuff. This simplifies the whole discussion.